Add `argument_and_relation_type_whitelist` parameter to `RETextClassificationWithIndicesTaskModule`

RainbowRivey opened the pull request:
This PR adds the parameter `argument_and_relation_type_whitelist` to `RETextClassificationWithIndicesTaskModule`. From the docstring:
```
argument_and_relation_type_whitelist: Union[Dict[str, List[List[str]]], List[List[str]]], optional, 
      defaults None. If set, only given relation types with given argument types will persist in 
      documents and generated by `create_relation_candidates`. This also affects predictions on 
      `decode()`, so it strictly filters both model input and output. Can also be passed as a list 
      of lists, where the first element is the relation type and the rest are the argument types.
```
E.g. when `argument_and_relation_type_whitelist={"org:founded_by: [['ORG', 'PER'],['ORG','ORG']]}` only relations with this particular types will remain in the document.

-  `_add_candidate_relations()` is also changed to respect this filter, so with above example, only candidates of types `{'no_relation': ['ORG', 'PER'], 'no_relation': ['ORG', 'ORG'] } will be added.
- `create_annotations_from_output()` also respects this filter.
- filtering occurs before `_add_reversed_relations()`, so reversed relations will be added only for relations that passed this filter.

NOTES:
- ~~There are currently no tests for `create_annotations_from_output()` for this taskmodule.~~ 

TODO:
 - [x] test in downstream setup (e.g. runs with 3 seeds at drugprot)

RainbowRivey committed:
Add `argument_and_relation_type_whitelist` parameter to `RETextClassificationWithIndicesTaskModule`

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -385,6 +385,7 @@ def __init__(
         add_global_attention_mask_to_input: bool = False,
         argument_type_whitelist: Optional[List[List[str]]] = None,
         handle_relations_with_same_arguments: str = "keep_none",
+        argument_and_relation_type_whitelist: Optional[Dict[str, List[str]]] = None,
```
I think the values should be `List[List[str]]`?

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -427,12 +428,19 @@ def __init__(
+        if argument_and_relation_type_whitelist is not None:
+            # hydra does not support tuples, so we got lists and need to convert them
+            self.argument_and_relation_type_whitelist = {
+                rel: (types[0], types[1])
```
the values should be a **Set of Tuples**

ArneBinder reviewed (changes requested):
Some issues below. Also: 
- Please use `get_relation_argument_spans_and_roles` everywhere where you check against the argument types so that we have just one single interface to get the argument types (you may change the original code in `_add_candidate_relations`). 
- Add two tests
  - `test_create_annotations_from_output` for the default case (default taskmodule parameters)
  - `test_create_annotations_from_output_with_argument_and_relation_type_whitelist` for the new functionality

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -427,12 +428,19 @@ def __init__(
             self.argument_type_whitelist = [
                 (types[0], types[1]) for types in argument_type_whitelist
             ]
+        if argument_and_relation_type_whitelist is not None:
```
After thinking about this a bit more, I think it is a good idea to do the handling of `argument_and_relation_type_whitelist` before `argument_type_whitelist` and if `argument_type_whitelist` is not provided, set it based on the values of `argument_and_relation_type_whitelist`. What do you think?

ArneBinder replied:
we could also do the other way around: remove entries not in ...and_relation_whitelist from the other

ArneBinder replied:
hmm ok, this seems to be not straight forward und hard to communicate to the user, let's not change one of the whitelists depending on the other. sorry for the noise

RainbowRivey committed:
Change `argument_and_relation_type_whitelist` items to be lists.

- Fix whitelist logic in `create_annotations_from_output`

- Add `test_create_annotations_from_output` + `test_create_annotations_from_output_with_argument_and_relation_type_whitelist`

ArneBinder commented on tests/taskmodules/test_re_text_classification_with_indices.py:
```diff
@@ -2882,3 +2964,86 @@ def test_encode_with_add_entity_tags_to_input_windowing(documents, insert_marker
+            ]
+        ),
+    }
+    task_encodings = taskmodule.encode(document, encode_target=True)
```
It is really hard to understand, what's happening here. On idea for improvement would be to show the candidate relations that will get classified:
```suggestion
    task_encodings = taskmodule.encode(document, encode_target=True)
    candidate_relations = [
        task_encoding.metadata["candidate_annotation"] for task_encoding in task_encodings
    ]
    resolved_candidate_relations = [rel.resolve() for rel in candidate_relations]
    assert resolved_candidate_relations == [
        ("per:employee_of", (("PER", "Entity G"), ("ORG", "H"))),
        ("per:founder", (("PER", "Entity G"), ("ORG", "I"))),
        ("org:founded_by", (("ORG", "I"), ("ORG", "H"))),
    ]
```

ArneBinder commented on tests/taskmodules/test_re_text_classification_with_indices.py:
```diff
@@ -2882,3 +2964,86 @@ def test_encode_with_add_entity_tags_to_input_windowing(documents, insert_marker
+
+
+def test_create_annotations_from_output(taskmodule, documents):
+    document = documents[4]
```
create the document from scratch, otherwise this test is hard to understand

ArneBinder commented on tests/taskmodules/test_re_text_classification_with_indices.py:
```diff
@@ -2882,3 +2964,86 @@ def test_encode_with_add_entity_tags_to_input_windowing(documents, insert_marker
+def test_create_annotations_from_output_with_argument_and_relation_type_whitelist(
+    documents, taskmodule
+):
+    document = documents[4]
```
create the document from scratch, otherwise this test is hard to understand

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -1336,7 +1388,21 @@ def create_annotations_from_output(
+            # Do not create annotation if
+            # - `add_candidate_relations` is True and relation label is a `none_label`
+            # - `argument_and_relation_type_whitelist` is used and a binary relation is not on a whitelist
+            if not (self.add_candidate_relations and label == self.none_label) and not (
+                isinstance(new_annotation, BinaryRelation)
+                and self.argument_and_relation_type_whitelist is not None
+                and not (
+                    label in self.argument_and_relation_type_whitelist
+                    and (
+                        getattr(new_annotation.head, "label"),
+                        getattr(new_annotation.tail, "label"),
+                    )
+                    in self.argument_and_relation_type_whitelist[label]
+                )
+            ):
```
This is super hard to read. Can you move the negations inside the brackets? Also, please use `get_relation_argument_spans_and_roles`.

EDIT: I think it is also fine to have nested `if`s if this is better to read.

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -427,12 +428,21 @@ def __init__(
         self.allow_discontinuous_text = allow_discontinuous_text
         self.handle_relations_with_same_arguments = handle_relations_with_same_arguments
         self.argument_type_whitelist: Optional[List[Tuple[str, str]]] = None
+        self.argument_and_relation_type_whitelist: Optional[Dict[str, List[Tuple[str, str]]]] = (
```
```suggestion
        self.argument_and_relation_type_whitelist: Optional[Dict[str, Set[Tuple[str, str]]]] = (
```

ArneBinder commented on src/pie_modules/taskmodules/re_text_classification_with_indices.py:
```diff
@@ -385,6 +385,7 @@ def __init__(
         add_global_attention_mask_to_input: bool = False,
         argument_type_whitelist: Optional[List[List[str]]] = None,
         handle_relations_with_same_arguments: str = "keep_none",
+        argument_and_relation_type_whitelist: Optional[Dict[str, List[List[str]]]] = None,
```
add documentation for the new parameter to the docstring above

ArneBinder reviewed (changes requested):
See below.

RainbowRivey committed:
Change intern argument type to tuple to support Nary relations.
+ add `no_relation` prediction case to `test_create_annotations_from_output`
+ improve tests
+ improve readability

ArneBinder committed:
improve readability of final if in create_annotations_from_output

ArneBinder committed:
cleanup and simplify test_create_annotations_from_output

ArneBinder commented on tests/taskmodules/test_re_text_classification_with_indices.py:
```diff
@@ -2882,3 +2964,186 @@ def test_encode_with_add_entity_tags_to_input_windowing(documents, insert_marker
+        ]
+
+
+def test_create_annotations_from_output_with_argument_and_relation_type_whitelist():
```
I simplified and removed all dependencies to other taskmodule methods in `test_create_annotations_from_output` and also parametrized `add_candidate_relations` (see https://github.com/ArneBinder/pie-modules/pull/170/commits/bedecd8777e2161988f8956db576f421856c0299). Please adjust `test_create_annotations_from_output_with_argument_and_relation_type_whitelist` accordingly.

RainbowRivey replied:
Changed the other test accordingly. I've added a few more relations and adjusted the filter to keep 'add_candidate_relations' effect visible.

ArneBinder replied:
cool, thanks!

ArneBinder committed:
fix missing document

ArneBinder reviewed (changes requested):
Please have a look. Also see [above](https://github.com/ArneBinder/pie-modules/pull/170#discussion_r2033578071).

RainbowRivey committed:
improve 'test_create_annotations_from_output_with_argument_and_relation_type_whitelist'

RainbowRivey commented:
We will probably need to change the argument type used to create TaskModule from
`argument_and_relation_type_whitelist: Optional[Dict[str, List[List[str]]]] = None` to `Optional[List[Union[str, List[List[str]]]]]`, since it is not a trivial problem to pass dict keys containing spaces to Hydra through command line 😥

ArneBinder committed:
allow to pass a list of lists as argument_and_relation_type_whitelist

ArneBinder committed:
adjust docstring

ArneBinder reviewed (approved):
I think this looks good now. If tested successfully downstream, this can be merged.
